polyval/src/backend/autodetect.rs

```diff
@@ -49,45 +55,41 @@ impl NewUniversalHash for Polyval {
     }
 }
 
-impl UniversalHash for Polyval {
+impl BlockSizeUser for Polyval {
     type BlockSize = U16;
+}
 
-    /// Input a field element `X` to be authenticated
-    #[inline]
-    fn update(&mut self, x: &Block) {
+impl ParBlocksSizeUser for Polyval {
+    type ParBlocksSize = U16;
+}
+
+impl UhfBackend for Polyval {
+    fn proc_block(&mut self, x: &Block) {
         if self.token.get() {
-            unsafe { (*self.inner.intrinsics).update(x) }
+            unsafe { (*self.inner.intrinsics).proc_block(x) }
         } else {
-            unsafe { (*self.inner.soft).update(x) }
+            unsafe { (*self.inner.soft).proc_block(x) }
```

Comment:
I will need some time to properly read changes in this PR, but the first thing which caught my eye is this part.

It looks like you misunderstood how the new API is supposed to be implemented. The idea is that `update_with_backend` passes to rank-2 closure `f` one of backends which was selected based on runtime detection of target features, thus "autodetection backend" should not exist. In other words, you should detect target features inside `update_with_backend` and call `f` with either CLMUL or software backend.

For reference, you can take a look at how ChaCha20 is [implemented](https://github.com/RustCrypto/stream-ciphers/blob/master/chacha20/src/lib.rs#L271-L282) (for now you can ignore the code for "force" features). The inner functions are used to [enable](https://github.com/RustCrypto/stream-ciphers/blob/master/chacha20/src/backends/avx2.rs#L19) target features checked earlier, which can significantly improve codegen for body of the rank-2 closure after it gets inlined. I understand that approach is not trivial, so feel free to ask questions about it.

I am not surprised that with the current implementation you got degraded performance, since you effectively check target feature on each processed block. Not only it introduces branching, but also prevents compiler from applying important optimizations.

Reply:
Indeed I did misunderstand the new traits; this makes much more sense. It would be helpful to document this on the traits themselves, as all I had to go off were the trait bounds, and the fact that `UhfBackend: ParBlocksSizeUser` and `ParBlocksSizeUser: BlockSizeUser` is what led me to believe that the top level struct should implement the backend.